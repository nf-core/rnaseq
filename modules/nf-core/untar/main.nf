process UNTAR {
    tag "$archive"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/ubuntu:20.04' :
        'nf-core/ubuntu:20.04' }"

    input:
    tuple val(meta), path(archive)

    output:
    tuple val(meta), path("$prefix"), emit: untar
    path "versions.yml"             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args  = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    prefix    = task.ext.prefix ?: ( meta.id ? "${meta.id}" : archive.baseName.toString().replaceFirst(/\.tar$/, ""))

    """
    mkdir $prefix

    ## Ensures --strip-components only applied when top level of tar contents is a directory
    ## If just files or multiple directories, place all in prefix
    if [[ \$(tar -taf ${archive} | grep -o -P "^.*?\\/" | uniq | wc -l) -eq 1 ]]; then
        tar \\
            -C $prefix --strip-components 1 \\
            -xavf \\
            $args \\
            $archive \\
            $args2
    else
        tar \\
            -C $prefix \\
            -xavf \\
            $args \\
            $archive \\
            $args2
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        untar: \$(echo \$(tar --version 2>&1) | sed 's/^.*(GNU tar) //; s/ Copyright.*\$//')
    END_VERSIONS
    """

    stub:
    prefix    = task.ext.prefix ?: ( meta.id ? "${meta.id}" : archive.toString().replaceFirst(/\.[^\.]+(.gz)?$/, ""))
    """
    mkdir $prefix
    touch ${prefix}/file.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        untar: \$(echo \$(tar --version 2>&1) | sed 's/^.*(GNU tar) //; s/ Copyright.*\$//')
    END_VERSIONS
    """
}
