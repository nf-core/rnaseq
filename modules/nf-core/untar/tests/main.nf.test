nextflow_process {

    name "Test Process UNTAR"
    script "../main.nf"
    process "UNTAR"

    tag "modules"
    tag "modules_nfcore"
    tag "untar"

    test("test_untar") {

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [ [], file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/db/kraken2.tar.gz', checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out.untar).match("test_untar") },
            )
        }

    }

    test("test_untar_onlyfiles") {

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [ [], file(params.modules_testdata_base_path + 'generic/tar/hello.tar.gz', checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out.untar).match("test_untar_onlyfiles") },
            )
        }

    }

}
