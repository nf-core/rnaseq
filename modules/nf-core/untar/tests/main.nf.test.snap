{
    "test_untar_onlyfiles": {
        "content": [
            [
                [
                    [
                        
                    ],
                    [
                        "hello.txt:md5,e59ff97941044f85df5297e1c302d260"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-10-18T11:56:46.878844"
    },
    "test_untar": {
        "content": [
            [
                [
                    [
                        
                    ],
                    [
                        "hash.k2d:md5,8b8598468f54a7087c203ad0190555d9",
                        "opts.k2d:md5,a033d00cf6759407010b21700938f543",
                        "taxo.k2d:md5,094d5891cdccf2f1468088855c214b2c"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-10-18T11:56:08.16574"
    }
}